netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+       */
+      if (bufferedWrites == null) {
+        super.write(ctx, msg, promise);
+      } else {
+        bufferedWrites.add(new ChannelWrite(msg, promise));
       }
+    }
 
-      @Override
-      public ListenableFuture<Void> completeFuture() {
-        return completionHandler.getUpgradeFuture();
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+      /**
+       * Don't propagate flush while there are still messages in the buffer.
+       * This is required to make buffering with the SslHandler work, as otherwise
+       * TLS handshakes "randomly" timeout. The SslHandler doesn't like flushes while the
+       * protocol is not negotiated...
+       */
+      if (bufferedWrites == null) {
+        ctx.flush();
       }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Exception {
+      failBufferedAndClose(ctx);
+    }
 
-      @Override
-      public void onConnected(Channel channel) {
-        // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
-        // which causes the upgrade headers to be added
-        DefaultHttpRequest upgradeTrigger =
-            new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
-        channel.writeAndFlush(upgradeTrigger);
+    void failBufferedAndClose(ChannelHandlerContext ctx) {
+      if (bufferedWrites != null) {
+        Exception e = new Exception("Buffered write failed.");
+        while (!bufferedWrites.isEmpty()) {
+          ChannelWrite write = bufferedWrites.poll();
+          write.promise.setFailure(e);
+        }
+        bufferedWrites = null;
       }
-    };
-  }
+      /**
+       * In case something goes wrong ensure that the channel gets closed as the
+       * NettyClientTransport relies on the channel's close future to get completed.
+       */
+      ctx.close();
+    }
 
-  /**
-   * Create a "no-op" negotiation that simply assumes the protocol to already be negotiated.
-   */
-  public static Negotiation plaintext(final ChannelHandler... handlers) {
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(Channel ch) throws Exception {
-        ch.pipeline().addLast(handlers);
+    void flushBufferedAndRemove(ChannelHandlerContext ctx) {
+      if (!ctx.channel().isActive() || flushed) {
+        return;
       }
-    };
-    return new Negotiation() {
-      private final SettableFuture<Void> completeFuture = SettableFuture.create();
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+      // Make sure that method can't be reentered, so that the ordering
+      // in the queue can't be messed up.
+      flushed = true;
+      while (!bufferedWrites.isEmpty()) {
+        ChannelWrite write = bufferedWrites.poll();
+        ctx.write(write.msg, write.promise);
       }
+      ctx.flush();
```

Comment:
@nmittler I think it's generally fine to flush once we emptied our buffer. For grpc we also know that we always need to flush.